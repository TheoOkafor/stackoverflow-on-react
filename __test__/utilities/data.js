const data = {
  answer: {
    accepted: null,
    body: 'What?',
    comments: [
      {
        answerid: 17,
        body: 'Jonzing',
        id: 30,
        questionid: 4,
        timesubmitted: '2018-11-21T12:35:08.000Z',
        userid: 4,
        username: 'teddy01'
      },
    ],
    downvotes: 0,
    id: 74,
    questionid: 38,
    timesubmitted: '2018-11-21T18:20:11.000Z',
    upvotes: 0,
    userid: 4,
    username: 'teddy01',
    votesCount: 0,
  },
  comment: {
    answerid: 17,
    body: 'Jonzing',
    id: 30,
    questionid: 4,
    timesubmitted: '2018-11-21T12:35:08.000Z',
    userid: 4,
    username: 'teddy01'
  },
  sampleAnswer: {
    data: {
      body: 'Jonzing',
      questionid: 4,
      timesubmitted: '2018-11-21T12:35:08.000Z',
      userid: 4,
      username: 'teddy01'
    },
  },
  question: {
    data: {
      answers: [
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 4,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 1,
          userid: 3,
          username: 'chanpion',
          votesCount: 1
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 8,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 12,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [],
          downvotes: 0,
          id: 16,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: false,
          body: 'Lorem ipsum dolor sit amet, ',
          comments: [{
            answerid: 17,
            body: 'Jonzing',
            id: 30,
            questionid: 4,
            timesubmitted: '2018-11-21T12:35:08.000Z',
            userid: 4,
            username: 'teddy01'
          }],
          downvotes: 0,
          id: 17,
          questionid: 4,
          timesubmitted: '2018-08-22T11:13:09.000Z',
          upvotes: 0,
          userid: 3,
          username: 'chanpion',
          votesCount: 0
        },
        {
          accepted: null,
          body: 'worddy',
          comments: [],
          downvotes: 0,
          id: 67,
          questionid: 4,
          timesubmitted: '2018-11-21T12:23:12.000Z',
          upvotes: 0,
          userid: 4,
          username: 'teddy01',
          votesCount: 0
        }
      ],
      body: 'Lorem ipsum dolor sit amet, ',
      id: 4,
      timesubmitted: '2018-08-22T11:13:09.000Z',
      title: 'How can one learn it all so fast?',
      userid: 3,
      username: 'chanpion'
    },
    message: 'Question 4 found',
    statusCode: 200
  },
  questions: [
    {
      id: 1,
      title: 'why do yo shake?',
      body: 'Lorem ipsum dolor sit amet, ',
      userid: 3,
      username: 'chanpion',
      answers: [],
    },
    {
      id: 2,
      title: 'why do yo shake?',
      body: 'Lorem ipsum dolor sit amet, ',
      userid: 3,
      username: 'chanpion',
      answers: [],
    },
    {
      id: 3,
      title: 'why do yo shake?',
      body: 'Lorem ipsum dolor sit amet, ',
      userid: 3,
      username: 'chanpion',
      answers: [],
    },
  ],
  newQuestion: {
    id: 1,
    title: 'why do yo shake?',
    body: 'Lorem ipsum dolor sit amet, ',
    userid: 3,
    username: 'chanpion',
    answers: [],
  },
  payload: {
    username: 'Xtreme',
    'x-access-token': 'token',
  },
  userData: {
    firstName: 'Xtreme',
    lastName: 'Prodigy',
    username: 'Xtreme',
    password: 'password',
    confirmpassword: 'password',
  },
};

export default data;
