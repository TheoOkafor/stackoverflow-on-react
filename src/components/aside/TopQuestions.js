import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchQuestions } from '../../actions/questionActions';
import hasAccepted from '../utilities/hasAccepted';

class TopQuestions extends Component {
  constructor(props) {
    super(props);

    this.state = {
      topQuestions: [],
    };
  }

  componentDidMount() {
    this.props.fetchQuestions();
  }

  render() {
    const topQuestions = [];
    if (this.props.questions.length > 0) {
      const { questions } = this.props;
      const questionSummary = [];
      const data = questions.slice(0).reverse();
      data.forEach((question) => {
        const accepted = hasAccepted(question.answers);
        questionSummary.push({
          id: question.id,
          numAnswers: question.numAnswers,
          title: question.title,
          hasAccepted: accepted,
        });
      });

      questionSummary.sort((a, b) => {
        return b.numAnswers - a.numAnswers;
      });
      for (let i = 0; i < 6; i += 1) {
        topQuestions.push(
          <tr key={questionSummary[i].id}>
            <td>
              <p
                className={
                  `num-answer
                ${questionSummary[i].hasAccepted ? 'selected' : ''}`}>
                { questionSummary[i].numAnswers }
              </p>
            </td>
            <td>
              <p><a href={`${
                window.location.href.split('/')[0]}/questions/${
                questionSummary[i].id}`}>
                { questionSummary[i].title }</a>
              </p>
            </td>
          </tr>
        );
      }
    }
    return (
      <table className="table" id="top-questions">
        <tbody>
          { topQuestions }
        </tbody>
      </table>
    );
  }
}

const mapStateToProps = state => ({
  questions: state.questions.questions,
});

TopQuestions.propTypes = {
  fetchQuestions: PropTypes.func.isRequired,
  questions: PropTypes.array,
};

export default connect(mapStateToProps, { fetchQuestions })(TopQuestions);
