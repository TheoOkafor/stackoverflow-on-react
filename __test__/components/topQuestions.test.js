import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import TopQuestions from '../../src/components/aside/TopQuestions';
import store from '../utilities/store';
import data from '../utilities/data';

const store1 = store({
  questions: {
    questions: [],
  }
});
const store2 = store(data.homeQuestions);

describe('AcceptAnswer component', () => {
  test('Test that the TopQuestions displays', () => {
    const component = renderer.create(
      <MemoryRouter>
        <TopQuestions store={store1}/>
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });

  test('Test that the TopQuestions displays', () => {
    const component = renderer.create(
      <MemoryRouter>
        <TopQuestions store={store2}/>
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });
});
