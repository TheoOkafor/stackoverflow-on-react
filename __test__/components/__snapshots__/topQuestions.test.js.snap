// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcceptAnswer component Test that the TopQuestions displays 1`] = `
<table
  className="table"
  id="top-questions"
>
  <tbody />
</table>
`;

exports[`AcceptAnswer component Test that the TopQuestions displays 2`] = `
<table
  className="table"
  id="top-questions"
>
  <tbody>
    <tr>
      <td>
        <p
          className="num-answer
                "
        />
      </td>
      <td>
        <p>
          <a
            href="http:/questions/6"
          >
            why do yo shake?
          </a>
        </p>
      </td>
    </tr>
    <tr>
      <td>
        <p
          className="num-answer
                "
        />
      </td>
      <td>
        <p>
          <a
            href="http:/questions/5"
          >
            why do yo shake?
          </a>
        </p>
      </td>
    </tr>
    <tr>
      <td>
        <p
          className="num-answer
                "
        />
      </td>
      <td>
        <p>
          <a
            href="http:/questions/4"
          >
            why do yo shake?
          </a>
        </p>
      </td>
    </tr>
    <tr>
      <td>
        <p
          className="num-answer
                "
        />
      </td>
      <td>
        <p>
          <a
            href="http:/questions/3"
          >
            why do yo shake?
          </a>
        </p>
      </td>
    </tr>
    <tr>
      <td>
        <p
          className="num-answer
                selected"
        >
          7
        </p>
      </td>
      <td>
        <p>
          <a
            href="http:/questions/2"
          >
            why do yo shake?
          </a>
        </p>
      </td>
    </tr>
    <tr>
      <td>
        <p
          className="num-answer
                "
        >
          7
        </p>
      </td>
      <td>
        <p>
          <a
            href="http:/questions/1"
          >
            why do yo shake?
          </a>
        </p>
      </td>
    </tr>
  </tbody>
</table>
`;
